server/validate.go

```diff
@@ -106,47 +106,61 @@ func validateGetEntryAndProofRequest(req *trillian.GetEntryAndProofRequest) erro
 }
 
 func validateQueueLeafRequest(req *trillian.QueueLeafRequest) error {
-	if req.Leaf == nil {
-		return status.Error(codes.InvalidArgument, "QueueLeafRequest.Leaf empty")
-	}
-	if err := validateLogLeaf(req.Leaf); err != nil {
-		// validateLogLeaf errors are meant to chain nicely with "Leaf."
-		return status.Errorf(codes.InvalidArgument, "QueueLeafRequest.Leaf.%v", err)
-	}
-	return nil
+	return validateLogLeaf(req.Leaf, "QueueLeafRequest.Leaf")
+}
+
+func validateAddSequencedLeafRequest(req *trillian.AddSequencedLeafRequest) error {
+	return validateLogLeaf(req.Leaf, "AddSequencedLeafRequest.Leaf")
 }
 
 func validateQueueLeavesRequest(req *trillian.QueueLeavesRequest) error {
-	if len(req.Leaves) == 0 {
-		return status.Error(codes.InvalidArgument, "QueueLeavesRequest.Leaves empty")
+	return validateLogLeaves(req.Leaves, "QueueLeavesRequest")
+}
+
+func validateAddSequencedLeavesRequest(req *trillian.AddSequencedLeavesRequest) error {
+	prefix := "AddSequencedLeavesRequest"
+	if err := validateLogLeaves(req.Leaves, prefix); err != nil {
+		return err
 	}
+
+	lastIndex := int64(-1)
 	for i, leaf := range req.Leaves {
-		if leaf == nil {
-			return status.Errorf(codes.InvalidArgument, "QueueLeavesRequest.Leaves[%v] empty", i)
-		}
-		if err := validateLogLeaf(leaf); err != nil {
-			// validateLogLeaf errors are meant to chain nicely with "Leaves."
-			return status.Errorf(codes.InvalidArgument, "QueueLeavesRequest.Leaves[%v].%v", i, err)
+		if leaf.LeafIndex <= lastIndex {
```

Comment:
Do we want to ensure the leaves are contiguous?

Reply:
Initially I wanted do so, but I was not sure.

Suppose you submit entries 0-99, but get a response that only entries 0,2,4,...,98 were saved. You can further submit entries 1,3,...,99 in a single batch if non-contiguous is allowed. If not, you can do one of:

extreme1: 50 one-element ranges (kills the advantage of batching)
extreme2: all 100 entries (50 entries are re-submitted)
trade-off between the two (e.g., [0-3] [5-8] ...)
I am not sure how much the above example is realistic.

Perhaps we can start with an API with tighter requirements (that is, submit only contiguous ranges). If it ends up too strict, the requirements can be loosened. WDYT?

Reply:
@gdbelvin In Key Transparency, do you ever need to submit arbitrary non-contiguous batches? Or single entry / contiguous batches would be sufficient?

Reply:
@AlCutter I propose to fine-tune this API in follow-up PRs, as I get more insight into what we can do with the batches in MySQL.